src/notebooks/controllers/notebookControllerManager.ts

```diff
@@ -457,12 +461,26 @@ export class NotebookControllerManager implements INotebookControllerManager, IE
             let preferredConnection: KernelConnectionMetadata | undefined;
             // Don't attempt preferred kernel search for interactive window, but do make sure we
             // load all our controllers for interactive window
+            const notebookMetadata = getNotebookMetadata(document);
             if (document.notebookType === JupyterNotebookView) {
-                preferredConnection = await this.kernelFinder.findKernel(
-                    document.uri,
-                    getNotebookMetadata(document),
-                    preferredSearchToken.token
-                );
+                ({ preferredConnection } = await this.findPreferredKernelExactMatch(
+                    document,
+                    preferredSearchToken.token,
+                    'useCache'
+                ));
+
+                // If we didn't find an exact match in the cache, try ignoring the cache
+                // IANHU - Slowdown? Only use cache for now
```

Comment:
Does this still work on first launch? I think the 'cached' version will do the real search if the cache is empty, but not 100% sure

Reply:
@rchiodo Still in draft form sadly :(. There are both tests failing and a massive slowdown in test execution. And I still have to consider returning top X results for suggestions if no exact match is found. I was just playing with this bit here. My guess was that it was failing to find an exact match more often (expected given stricter matching) so way more often it was having to wait for full non-cached results.

Not sure on first launch.

Reply:
VS Code test suites went from like 18 min => 60+ min.